### PR TITLE
[cmake] Run src/ tests

### PR DESCRIPTION
**Pull request by ebraminio**

### CODE REVIEW

**Comment by ebraminio** [comment]
The command for fully testing src/ folder is 
```
cmake -DBUILD_SHARED_LIBS=ON -DHB_HAVE_FREETYPE=ON -DHB_HAVE_GRAPHITE2=ON -DHB_BUILTIN_UCDN=ON -DHB_HAVE_GLIB=ON -DHB_BUILD_UTILS=ON -Bbuild -H. -GNinja && ninja -Cbuild && CTEST_OUTPUT_ON_FAILURE=1 CTEST_PARALLEL_LEVEL=8 ninja -Cbuild test
```

but this doesn't work very well on Mac at least.

File: src/check-symbols.sh [hunk 1]
```diff
@@ -17,10 +18,10 @@ fi
 echo "Checking that we are not exposing internal symbols"
 tested=false
 for suffix in so dylib; do
-	so=.libs/libharfbuzz.$suffix
+	so=$libs/libharfbuzz.$suffix
 	if ! test -f "$so"; then continue; fi
 
-	EXPORTED_SYMBOLS="`nm "$so" | grep ' [BCDGINRSTVW] .' | grep -v ' _fini\>\| _init\>\| _fdata\>\| _ftext\>\| _fbss\>\| __bss_start\>\| __bss_start__\>\| __bss_end__\>\| _edata\>\| _end\>\| _bss_end__\>\| __end__\>\| __gcov_flush\>\| ___gcov_flush\>\| llvm_\| _llvm_' | cut -d' ' -f3`"
+	EXPORTED_SYMBOLS="`nm "$so" | grep ' [BCDGINRSTVW] .' | grep -v ' _fini\>\| _init\>\| _fdata\>\| _ftext\>\| _fbss\>\| __bss_start\>\| __bss_start__\>\| __bss_end__\>\| _edata\>\| _end\>\| _bss_end__\>\| __end__\>\| __gcov_flush\>\| ___gcov_flush\>\| llvm_\| _llvm_\| __ZN' | cut -d' ' -f3`"
```
Comment (ebraminio) [inline src/check-symbols.sh]: Is exporting __ZN... symbols normal? I see things like 

```
__ZN2OT4CBDT13accelerator_t4initEP9hb_face_t
__ZN2OT4GPOS14position_startEP9hb_font_tP11hb_buffer_t
__ZN2OT4GPOS23position_finish_offsetsEP9hb_font_tP11hb_buffer_t
__ZN2OT4GPOS24position_finish_advancesEP9hb_font_tP11hb_buffer_t
__ZN2OT4GSUB16substitute_startEP9hb_font_tP11hb_buffer_t
__ZN2OT4cmap13accelerator_t14get_glyph_fromINS_12CmapSubtableEEEbPKvjPj
__ZN2OT4cmap13accelerator_t14get_glyph_fromINS_20CmapSubtableFormat12EEEbPKvjPj
__ZN2OT4cmap13accelerator_t21get_glyph_from_symbolINS_12CmapSubtableEEEbPKvjPj
__ZN2OT4cmap13accelerator_t4finiEv
__ZN2OT4cmap13accelerator_t4initEP9hb_face_t
__ZN2OT4glyf13accelerator_t4finiEv
__ZN2OT4glyf13accelerator_t4initEP9hb_face_t
__ZN2OT4kern13accelerator_t4finiEv
__ZN2OT4kern13accelerator_t4initEP9hb_face_t
__ZN2OT4post13accelerator_t4finiEv
__ZN2OT4post13accelerator_t4initEP9hb_face_t
__ZN2OT4post13accelerator_t7cmp_keyEPKvS3_Pv
__ZN2OT4post13accelerator_t8cmp_gidsEPKvS3_Pv
__ZN2OT5BEIntIjLi4EE3setEj
__ZN2OT5BEIntIsLi2EE3setEs
__ZN2OT5BEIntItLi2EE3setEt
```
...

**Comment by ebraminio** [comment]
Are these correct catches from check-static-inits?

```
Start  1: check-static-inits.sh
 1/60 Test  #1: check-static-inits.sh ......................***Failed    2.07 sec
.
Checking that no object file has static initializers
Checking that no object file has lazy static C++ constructors/destructors or other such stuff
0000000000000000         *UND*	___cxa_guard_acquire
0000000000000000         *UND*	___cxa_guard_release
Ouch, ./hb-blob.cc.o has lazy static C++ constructors/destructors or other such stuff
0000000000000000         *UND*	___cxa_guard_acquire
0000000000000000         *UND*	___cxa_guard_release
Ouch, ./hb-buffer.cc.o has lazy static C++ constructors/destructors or other such stuff
0000000000000000         *UND*	___cxa_begin_catch
Ouch, ./hb-coretext.cc.o has lazy static C++ constructors/destructors or other such stuff
0000000000000000         *UND*	___cxa_guard_acquire
0000000000000000         *UND*	___cxa_guard_release
Ouch, ./hb-font.cc.o has lazy static C++ constructors/destructors or other such stuff
0000000000000000         *UND*	___cxa_begin_catch
Ouch, ./hb-ot-layout.cc.o has lazy static C++ constructors/destructors or other such stuff
```

**Comment by ebraminio** [comment]
CMake seem to export more than needed and the tests seem to broken. I see things like 

```
__ZN2OT4CBDT13accelerator_t4initEP9hb_face_t
__ZN2OT4GPOS14position_startEP9hb_font_tP11hb_buffer_t
__ZN2OT4GPOS23position_finish_offsetsEP9hb_font_tP11hb_buffer_t
__ZN2OT4GPOS24position_finish_advancesEP9hb_font_tP11hb_buffer_t
__ZN2OT4GSUB16substitute_startEP9hb_font_tP11hb_buffer_t
__ZN2OT4cmap13accelerator_t14get_glyph_fromINS_12CmapSubtableEEEbPKvjPj
__ZN2OT4cmap13accelerator_t14get_glyph_fromINS_20CmapSubtableFormat12EEEbPKvjPj
__ZN2OT4cmap13accelerator_t21get_glyph_from_symbolINS_12CmapSubtableEEEbPKvjPj
__ZN2OT4cmap13accelerator_t4finiEv
__ZN2OT4cmap13accelerator_t4initEP9hb_face_t
__ZN2OT4glyf13accelerator_t4finiEv
__ZN2OT4glyf13accelerator_t4initEP9hb_face_t
__ZN2OT4kern13accelerator_t4finiEv
__ZN2OT4kern13accelerator_t4initEP9hb_face_t
__ZN2OT4post13accelerator_t4finiEv
__ZN2OT4post13accelerator_t4initEP9hb_face_t
__ZN2OT4post13accelerator_t7cmp_keyEPKvS3_Pv
__ZN2OT4post13accelerator_t8cmp_gidsEPKvS3_Pv
__ZN2OT5BEIntIjLi4EE3setEj
__ZN2OT5BEIntIsLi2EE3setEs
__ZN2OT5BEIntItLi2EE3setEt
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -696,6 +699,32 @@ if (NOT SKIP_INSTALL_LIBRARIES AND NOT SKIP_INSTALL_ALL)
   endif ()
 endif ()
 
+
+## Tests
+if (UNIX)
+  if (BUILD_SHARED_LIBS)
+    add_test(NAME check-static-inits.sh
+      COMMAND ${PROJECT_SOURCE_DIR}/src/check-static-inits.sh
+      WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/CMakeFiles/harfbuzz.dir/src # hack
```
Comment (ebraminio) [inline CMakeLists.txt]: This is essentially a hack, don't know of a better way

File: CMakeLists.txt [hunk 3]
```diff
@@ -696,6 +699,32 @@ if (NOT SKIP_INSTALL_LIBRARIES AND NOT SKIP_INSTALL_ALL)
...
+    # does some "make" stuff inside
+    # add_test(NAME check-defs.sh COMMAND ${PROJECT_SOURCE_DIR}/src/check-defs.sh)
+    add_test(NAME check-libstdc++.sh COMMAND ${PROJECT_SOURCE_DIR}/src/check-libstdc++.sh)
+    add_test(NAME check-symbols.sh COMMAND ${PROJECT_SOURCE_DIR}/src/check-symbols.sh)
+  else ()
+    message(WARN "You can enable BUILD_SHARED_LIBS for more tests.")
+  endif ()
+
+  # disabled as it doesn't work anyway
+  if (OFF AND IN_HB_DIST)
```
Comment (ebraminio) [inline CMakeLists.txt]: These really should be enabled sometime...

**Comment by ebraminio** [comment]
imperfect, but lets have it so others can hack upon it.

**Comment by waywardmonkeys** [comment]
Would it make sense to file a separate issue for the various things you're mentioning here? I don't mind trying to help look at them, but clear issues would make that easier for me. :)

**Comment by ebraminio** [comment]
#678, feel free to have a look at any of them you like :)

File: src/check-defs.sh [hunk 4]
```diff
@@ -4,6 +4,7 @@ LC_ALL=C
 export LC_ALL
 
 test -z "$srcdir" && srcdir=.
+test -z "$libs" && libs=.
```
Comment (behdad) [inline src/check-defs.sh]: Humm. I prefer libs to default to .libs

Comment (behdad) [reply src/check-defs.sh]: Or maybe .libs if that dir exists, otherwise .

Comment (ebraminio) [reply src/check-defs.sh]: On #679

**Pull request merged**